matchzoo/datapack.py

```diff
@@ -0,0 +1,114 @@
+"""Matchzoo DataPack, pair-wise tuple (feature) and context as input."""
+
+import typing
+from pathlib import Path
+
+import dill
+import pandas as pd
+
+
+class DataPack(pd.DataFrame):
+    """
+    Matchzoo DataPack data structure, store dataframe and context.
+
+    Example:
+        >>> # features, context generate by processors.
+        >>> features = [([1,3], [2,3]), ([3,0], [1,6])]
+        >>> context = {'vocab_size': 2000}
+        >>> dp = DataPack(data=features,
+        ...               context=context)
+        >>> dp.context
+        {'vocab_size': 2000}
+        >>> # sample without replacement for generation.
+        >>> type(dp.sample(1))
+        <class 'matchzoo.datapack.DataPack'>
+        >>> dp.size
+        2
+        >>> features, context = dp.unpack()
+    """
+
+    _metadata = ['context']
+
+    DATA_FILENAME = 'data.pkl'
+    CONTEXT_FILENAME = 'context.pkl'
+
+    def __init__(self,
+                 data: list,
+                 context: dict={},
+                 index: list= None,
+                 columns: list=['text_left', 'text_right'],
+                 dtype: object=None,
+                 copy: bool=True):
+        """Initialize."""
+        super(DataPack, self).__init__(data=data,
+                                       index=index,
+                                       columns=columns,
+                                       dtype=dtype,
+                                       copy=copy)
+        if self.shape[1] != 2:
+            raise ValueError(
+                "Pair-wise input expected.")
+        self.context = context
+
+    @property
+    def _constructor(self) -> callable:
+        """Subclass pd.DataFrame."""
+        return DataPack._internal_ctor
+
+    @classmethod
+    def _internal_ctor(cls, *args, **kwargs):
+        """Create subclass inputs to store context."""
+        kwargs['context'] = None
+        return cls(*args, **kwargs)
+
+    @property
+    def size(self) -> int:
+        """Get size of the data pack."""
+        return self.shape[0]
+
+    def unpack(self) -> typing.Union[pd.DataFrame, dict]:
```

Comment:
Redundant method. When you have `self`, you also have `self.context`.

Reply:
their is some redundancy here. taken into consider the user might not aware the `context` variable, I believe `unpack` is more natural method to access the data.

To some extent, i think sacrifice is worthwhile.

Reply:
I don't think so. Letting users know there's a `context` property takes same effort as letting users know there's a `unpack` method. And using `unpack` doesn't even save a single line of code: `features, context = dp.unpack()` is exactly the same as `features, context = dp, dp.context`. It even uses one more line since when the user know there's a `context` property, the user won't need to unpack or something, just directly access it when needed. Plus the type hint is wrong, it should be a `tuple` type. Plus the type hint is a bit confusing: yes, you're returning a `pd.DataFrame` since it's a subclass, but `DataPack` servers very different purposes and should be mentioned explicitly.